File: MP8V2_SparkSQL_Template_2023_V2_Testing/python/MP8_PartF.py
from pyspark.sql.functions import col, lag
from pyspark import SparkContext
from pyspark.sql.types import StructType
from pyspark.sql.types import StructField
from pyspark.sql.types import StringType, IntegerType
from pyspark.sql import SparkSession
from pyspark.sql.window import Window
from pyspark.sql.functions import *

sc = SparkContext()
spark = SparkSession.builder.getOrCreate()

####
# 1. Setup : Write a function to load it in an RDD & DataFrame
####

# RDD API
# Columns:
# 0: word (string), 1: year (int), 2: frequency (int), 3: books (int)
rdd = sc.textFile("gbooks")

schema = StructType([
    StructField("word", StringType(), True),
    StructField("year", StringType(), True),
    StructField("frequency", StringType(), True),
    StructField("books", StringType(), True)
])

df = spark.createDataFrame(rdd.map(lambda x: x.split("\t")), schema)

df.createOrReplaceTempView("MP8")

###
# 2. Frequency Increase (16 points): analyze the frequency increase of words starting from the year 1500 to the year 2000
###
# Spark SQL - DataFrame API
df_filtered = df.filter((col("year") >= 1500) & (col("year") <= 2000))
window = Window.partitionBy("word").orderBy("year")
df_with_increase = df_filtered.withColumn("frequency_increase", lag("frequency", -1).over(window).cast(IntegerType()))
df_with_increase = df_with_increase.groupBy("word").agg(sum("frequency_increase").alias("total_increase"))
df_with_increase = df_with_increase.orderBy(desc("total_increase"))
df_with_increase.show(20)

File: MP8V2_SparkSQL_Template_2023_V2_Testing/java/src/main/java/MP8_PartF.java
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.sql.*;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.sql.expressions.Window;
import org.apache.spark.sql.expressions.WindowSpec;
import static org.apache.spark.sql.functions.*;

public class MP8_PartF {
    public static void main(String[] args) throws IOException {
        SparkSession spark = SparkSession
        .builder()
        .appName("MP8")
        .getOrCreate();
        JavaSparkContext sc = new JavaSparkContext(spark.sparkContext());
        SQLContext sqlContext = new SQLContext(sc);
        /*
        * 1. Setup: write a function to load it in an RDD & DataFrame
        */
        
        // RDD API
        // Columns: 0: word (string), 1: year (int), 2: frequency (int), 3: books (int)


        /**
         * 2. Frequency Increase (16 points): analyze the frequency increase of words starting from the year 1500 to the year 2000
         */
        
    }
}